src/schemes/boundary/open_boundary/system.jl

```diff
@@ -0,0 +1,476 @@
+
+    print(io, "OpenBoundarySPHSystem{", ndims(system), "}(")
+    print(io, type2string(system.boundary_zone))
+    print(io, ") with ", nparticles(system), " particles")
+end
+
+function Base.show(io::IO, ::MIME"text/plain", system::OpenBoundarySPHSystem)
+    @nospecialize system # reduce precompilation time
+
+    if get(io, :compact, false)
+        show(io, system)
+    else
+        summary_header(io, "OpenBoundarySPHSystem{$(ndims(system))}")
+        summary_line(io, "#particles", nparticles(system))
+        summary_line(io, "#buffer_particles", system.buffer.buffer_size)
+        summary_line(io, "fluid system", type2string(system.fluid_system))
+        summary_line(io, "boundary", type2string(system.boundary_zone))
+        summary_line(io, "flow direction", system.flow_direction)
+        summary_line(io, "prescribed velocity", string(nameof(system.reference_velocity)))
+        summary_line(io, "prescribed pressure", string(nameof(system.reference_pressure)))
+        summary_line(io, "prescribed density", string(nameof(system.reference_density)))
+        summary_line(io, "width", round(system.boundary_zone.zone_width, digits=3))
+        summary_footer(io)
+    end
+end
+
+update_callback_used!(system::OpenBoundarySPHSystem) = system.update_callback_used[] = true
+
+@inline source_terms(system::OpenBoundarySPHSystem) = nothing
+
+@inline hydrodynamic_mass(system::OpenBoundarySPHSystem, particle) = system.mass[particle]
+
+@inline function particle_density(v, system::OpenBoundarySPHSystem, particle)
+    return system.density[particle]
+end
+
+@inline function particle_pressure(v, system::OpenBoundarySPHSystem, particle)
+    return system.pressure[particle]
+end
+
+function update_final!(system::OpenBoundarySPHSystem, v, u, v_ode, u_ode, semi, t;
+                       update_from_callback=false)
+    if !update_from_callback && !(system.update_callback_used[])
+        throw(ArgumentError("`UpdateCallback` is required when using `OpenBoundarySPHSystem`"))
+    end
+
+    @trixi_timeit timer() "evaluate characteristics" evaluate_characteristics!(system, v, u,
+                                                                               v_ode, u_ode,
+                                                                               semi, t)
+end
+
+# This function is called by the `UpdateCallback`, as the integrator array might be modified
+function update_open_boundary_eachstep!(system::OpenBoundarySPHSystem, v_ode, u_ode,
+                                        semi, t)
+    u = wrap_u(u_ode, system, semi)
+    v = wrap_v(v_ode, system, semi)
+
+    # Update density, pressure and velocity depending on the characteristic variables.
+    # See eq. 13-15 in Lastiwka (2009) https://doi.org/10.1002/fld.1971
+    @trixi_timeit timer() "update quantities" update_quantities!(system, v, u, t)
+
+    @trixi_timeit timer() "check domain" check_domain!(system, v, u, v_ode, u_ode, semi)
```

Comment:
I think the name "check domain" is not very descriptive. But I also don't have any better ideas...